File: src/lam.ts
export type Term
  = { $: "Var"; name: string }
  | { $: "Ref"; name: string }
  | { $: "Lam"; name: string; body: (x: Term) => Term }
  | { $: "App"; func: Term; argm: Term };

export type Book = { [name: string]: Term };
export type Stats = { beta: number };

export function Var(name: string): Term {
  return { $: "Var", name };
}

export function Ref(name: string): Term {
  return { $: "Ref", name };
}

export function Lam(name: string, body: (x: Term) => Term): Term {
  return { $: "Lam", name, body };
}

export function App(func: Term, argm: Term): Term {
  return { $: "App", func, argm };
}

// Names

export function name_of(n: number): string {
  if (n < 26) {
    return String.fromCharCode(97 + n);
  }
  return name_of(Math.floor(n / 26) - 1) + String.fromCharCode(97 + n % 26);
}

// Show

type Unapplied = { fun: Term; args: Term[] };

function unapp(term: Term): Unapplied {
  if (term.$ !== "App") {
    return { fun: term, args: [] };
  }
  var sp = unapp(term.func);
  return { fun: sp.fun, args: [...sp.args, term.argm] };
}

function show_head(term: Term, d: number): string {
  if (term.$ === "Lam") {
    return "(" + show_go(term, d) + ")";
  }
  return show_go(term, d);
}

function show_lam(term: Term & { $: "Lam" }, d: number): string {
  var name = name_of(d);
  return "λ" + name + "." + show_go(term.body(Var(name)), d + 1);
}

function show_app(term: Term & { $: "App" }, d: number): string {
  var sp = unapp(term);
  return show_head(sp.fun, d) + "(" + sp.args.map(a => show_go(a, d)).join(", ") + ")";
}

function show_go(term: Term, d: number): string {
  switch (term.$) {
    case "Var": return term.name;
    case "Ref": return "@" + term.name;
    case "Lam": return show_lam(term, d);
    case "App": return show_app(term, d);
  }
}

export function show(term: Term): string {
  return show_go(term, 0);
}

// Parse

type Bind = [string, Term];
type P = (ctx: Bind[]) => Term;

export function parse(code: string): Book {
  var i = 0;

  function skip() {
    while (i < code.length) {
      if (code[i] === '/' && code[i+1] === '/') {
        while (i < code.length && code[i] !== '\n') i++;
      } else if (/[ \n]/.test(code[i])) {
        i++;
      } else {
        break;
      }
    }
  }

  function parse_char(c: string) {
    if (code[i++] !== c) throw "Expected '" + c + "'";
  }

  function parse_name(): string {
    var s = "";
    while (i < code.length && /[0-9A-Za-z_]/.test(code[i])) s += code[i++];
    return s;
  }

  function parse_calls(fn: P): P {
    if (code[i] !== "(") {
      return fn;
    }
    i++;
    var args: P[] = [];
    while (true) {
      skip();
      if (code[i] === ")") {
        i++;
        break;
      }
      if (args.length > 0) {
        parse_char(",");
      }
      args.push(parse_term());
    }
    return parse_calls(ctx => args.reduce((f, a) => App(f, a(ctx)), fn(ctx)));
  }

  function parse_term(): P {
    skip();
    switch (code[i]) {
      case "λ": {
        i++;
        var n = parse_name();
        parse_char(".");
        var body = parse_term();
        return ctx => Lam(n, x => body([[n, x], ...ctx]));
      }
      case "@": {
        i++;
        var n = parse_name();
        return parse_calls(_ => Ref(n));
      }
      case "(": {
        i++;
        var inner = parse_term();
        parse_char(")");
        return parse_calls(inner);
      }
      default: {
        var n = parse_name();
        return parse_calls(ctx => ctx.find(b => b[0] === n)![1]);
      }
    }
  }

  var book: Book = {};
  while (true) {
    skip();
    if (i >= code.length) {
      break;
    }
    parse_char("@");
    var n = parse_name();
    skip();
    parse_char("=");
    book[n] = parse_term()([]);
    skip();
  }
  return book;
}

// Binary

function nat_to_bin(n: number): string {
  if (!Number.isInteger(n) || n < 0) {
    throw new Error("Invalid natural number: " + n);
  }
  return "1".repeat(n) + "0";
}

function bits_per_ref(defs: string[]): number {
  return Math.ceil(Math.log2(Math.max(defs.length, 1)));
}

function index_to_bits(index: number, width: number): string {
  if (!Number.isInteger(index) || index < 0 || index >= 2 ** width) {
    throw new Error("Invalid global reference index: " + index);
  }
  return index.toString(2).padStart(width, "0");
}

function bits_to_index(bits: string, pos: { i: number }, width: number): number {
  if (width === 0) return 0;
  var end = pos.i + width;
  if (end > bits.length) throw new Error("Invalid binary");
  var chunk = bits.slice(pos.i, end);
  if (!/^[01]+$/.test(chunk)) throw new Error("Invalid binary");
  pos.i = end;
  return parseInt(chunk, 2);
}

function term_to_bin(term: Term, d: number, defs: string[]): string {
  switch (term.$) {
    case "Var": return "10" + nat_to_bin(d - parseInt(term.name.slice(1)) - 1);
    case "Ref": return "11" + ref_to_bin(term.name, defs);
    case "Lam": return "00" + term_to_bin(term.body(Var("$" + d)), d + 1, defs);
    case "App": return "01" + bin_func(term, d, defs) + bin_argm(term, d, defs);
  }
}

function ref_to_bin(name: string, defs: string[]): string {
  var index = defs.indexOf(name);
  if (index < 0) throw new Error("Undefined reference: @" + name);
  return index_to_bits(index, bits_per_ref(defs));
}

function bin_func(term: Term & { $: "App" }, d: number, defs: string[]): string {
  return term_to_bin(term.func, d, defs);
}

function bin_argm(term: Term & { $: "App" }, d: number, defs: string[]): string {
  return term_to_bin(term.argm, d, defs);
}

export function to_bin(book: Book): string {
  var defs = Object.keys(book);
  var bits = nat_to_bin(defs.length);
  for (var i = 0; i < defs.length; i++) {
    bits += term_to_bin(book[defs[i]], 0, defs);
  }
  return bits;
}

function read_nat(bits: string, pos: { i: number }): number {
  var n = 0;
  while (bits[pos.i] === "1") {
    n++;
    pos.i++;
  }
  if (bits[pos.i] !== "0") throw new Error("Invalid binary");
  pos.i++;
  return n;
}

function read_term(bits: string, pos: { i: number }, d: number, dns: string[]): string {
  var tag = bits[pos.i++]! + bits[pos.i++]!;
  switch (tag) {
    case "10": return name_of(d - read_nat(bits, pos) - 1);
    case "11": return read_ref(bits, pos, dns);
    case "00": return "λ" + name_of(d) + "." + read_term(bits, pos, d + 1, dns);
    case "01": return read_app(bits, pos, d, dns);
    default: throw "Invalid binary";
  }
}

function read_ref(bits: string, pos: { i: number }, dns: string[]): string {
  var index = bits_to_index(bits, pos, bits_per_ref(dns));
  if (index >= dns.length) throw new Error("Invalid binary");
  return "@" + dns[index];
}

function read_app(bits: string, pos: { i: number }, d: number, dns: string[]): string {
  var func = read_term(bits, pos, d, dns);
  var argm = read_term(bits, pos, d, dns);
  if (func.startsWith("λ")) {
    return "(" + func + ")(" + argm + ")";
  }
  return func + "(" + argm + ")";
}

export function from_bin(bits: string): string {
  var pos = { i: 0 };
  var count = read_nat(bits, pos);
  var dns = Array.from({ length: count }, (_, i) => name_of(i));
  var lines: string[] = [];
  for (var idx = 0; idx < count; idx++) {
    lines.push("@" + dns[idx] + " = " + read_term(bits, pos, 0, dns));
  }
  if (pos.i !== bits.length) throw new Error("Invalid binary");
  return lines.join("\n") + "\n";
}

// Affinity check

type Uses = { [key: string]: number };

function merge_uses(a: Uses, b: Uses): Uses {
  var result: Uses = { ...a };
  for (var k in b) {
    result[k] = (result[k] || 0) + b[k];
  }
  return result;
}

function check_lam(term: Term & { $: "Lam" }, d: number, dn: string, errs: string[]): Uses {
  var vname = "$" + d;
  var uses = check_term(term.body(Var(vname)), d + 1, dn, errs);
  var count = uses[vname] || 0;
  if (count > 1) {
    errs.push("@" + dn + ": λ" + name_of(d) + " used " + count + " times");
  }
  delete uses[vname];
  return uses;
}

function check_app(term: Term & { $: "App" }, d: number, dn: string, errs: string[]): Uses {
  return merge_uses(
    check_term(term.func, d, dn, errs),
    check_term(term.argm, d, dn, errs));
}

function check_term(term: Term, d: number, dn: string, errs: string[]): Uses {
  switch (term.$) {
    case "Var": return { [term.name]: 1 };
    case "Ref": return {};
    case "Lam": return check_lam(term, d, dn, errs);
    case "App": return check_app(term, d, dn, errs);
  }
}

export function check_affine(book: Book): string[] {
  var errs: string[] = [];
  for (var name in book) {
    check_term(book[name], 0, name, errs);
  }
  return errs;
}

// Normalize

function wnf_ref(book: Book, name: string): Term {
  if (!(name in book)) {
    return Ref(name);
  }
  return book[name];
}

function wnf_app(book: Book, func: Term, argm: Term, stats: Stats): Term {
  if (func.$ === "Lam") {
    stats.beta++;
    return wnf(book, func.body(argm), stats);
  }
  return App(func, argm);
}

export function wnf(book: Book, term: Term, stats: Stats): Term {
  switch (term.$) {
    case "Var": return term;
    case "Ref": return wnf(book, wnf_ref(book, term.name), stats);
    case "Lam": return term;
    case "App": return wnf_app(book, wnf(book, term.func, stats), term.argm, stats);
  }
}

export function snf(book: Book, term: Term, stats: Stats): Term {
  var t = wnf(book, term, stats);
  switch (t.$) {
    case "Var": return t;
    case "Ref": return t;
    case "Lam": return Lam(t.name, x => snf(book, t.body(x), stats));
    case "App": return App(snf(book, t.func, stats), snf(book, t.argm, stats));
  }
}

File: hs/lam.hs
module Lam
  ( Book
  , Term(..)
  , parseBook
  , showTerm
  , snf
  , toBin
  , fromBin
  , checkAffine
  ) where

import Data.Char (isAlphaNum, isSpace)
import Data.List (findIndex, intercalate, isPrefixOf)

data Term
  = Var String
  | Ref String
  | Lam String (Term -> Term)
  | App Term Term

type Book = [(String, Term)]
type Env = [(String, Term)]
type Uses = [(String, Int)]

nameOf :: Int -> String
nameOf n
  | n < 26 = [toEnum (97 + n)]
  | otherwise = nameOf (n `div` 26 - 1) ++ [toEnum (97 + n `mod` 26)]

unapp :: Term -> (Term, [Term])
unapp (App f x) = let (h, xs) = unapp f in (h, xs ++ [x])
unapp term = (term, [])

showTerm :: Term -> String
showTerm term = go 0 term where
  go _ (Var n) = n
  go _ (Ref n) = "@" ++ n
  go d (Lam _ body) =
    let n = nameOf d in "\955" ++ n ++ "." ++ go (d + 1) (body (Var n))
  go d app@(App _ _) =
    let (h, xs) = unapp app
    in showHead d h ++ "(" ++ intercalate ", " (map (go d) xs) ++ ")"
  showHead d lam@(Lam _ _) = "(" ++ go d lam ++ ")"
  showHead d other = go d other

newtype Parser a = Parser { runParser :: String -> Either String (a, String) }

instance Functor Parser where
  fmap f p = Parser $ \s -> do
    (x, s') <- runParser p s
    pure (f x, s')

instance Applicative Parser where
  pure x = Parser $ \s -> Right (x, s)
  pf <*> px = Parser $ \s -> do
    (f, s') <- runParser pf s
    (x, s'') <- runParser px s'
    pure (f x, s'')

instance Monad Parser where
  p >>= f = Parser $ \s -> do
    (x, s') <- runParser p s
    runParser (f x) s'

skip :: Parser ()
skip = Parser $ \s -> Right ((), go s) where
  go ('/':'/':xs) = go (drop 1 (dropWhile (/= '\n') xs))
  go (x:xs) | isSpace x = go xs
  go xs = xs

peek :: Parser (Maybe Char)
peek = Parser $ \s -> Right (case s of [] -> Nothing; c:_ -> Just c, s)

char :: Char -> Parser ()
char c = Parser $ \s -> case s of
  x:xs | x == c -> Right ((), xs)
  _ -> Left ("Expected '" ++ [c] ++ "'")

name :: Parser String
name = Parser $ \s ->
  let (n, rest) = span isNameChar s
  in if null n then Left "Expected name" else Right (n, rest)
  where
    isNameChar x = isAlphaNum x || x == '_'

parseCalls :: Parser (Env -> Term) -> Parser (Env -> Term)
parseCalls pf = do
  skip
  m <- peek
  case m of
    Just '(' -> do
      char '('
      args <- parseArgs []
      parseCalls $ do
        f <- pf
        pure $ \env -> foldl App (f env) [a env | a <- args]
    _ -> pf

parseArgs :: [Env -> Term] -> Parser [Env -> Term]
parseArgs args = do
  skip
  m <- peek
  case m of
    Just ')' -> char ')' >> pure args
    _ -> do
      if null args then pure () else char ','
      arg <- parseTerm
      parseArgs (args ++ [arg])

parseTerm :: Parser (Env -> Term)
parseTerm = do
  skip
  m <- peek
  case m of
    Just '\955' -> do
      char '\955'
      n <- name
      char '.'
      body <- parseTerm
      pure $ \env -> Lam n $ \x -> body ((n, x) : env)
    Just '@' -> do
      char '@'
      n <- name
      parseCalls (pure $ const (Ref n))
    Just '(' -> do
      char '('
      inner <- parseTerm
      skip
      char ')'
      parseCalls (pure inner)
    _ -> do
      n <- name
      parseCalls $ pure $ \env -> case lookup n env of
        Just v -> v
        Nothing -> error ("unbound variable: " ++ n)

parseDef :: Parser (String, Term)
parseDef = do
  skip
  char '@'
  n <- name
  skip
  char '='
  term <- parseTerm
  pure (n, term [])

parseDefs :: Book -> Parser Book
parseDefs book = do
  skip
  m <- peek
  case m of
    Nothing -> pure book
    _ -> do
      def <- parseDef
      parseDefs (book ++ [def])

parseBook :: String -> Either String Book
parseBook code = do
  (book, rest) <- runParser (parseDefs []) code
  let rest' = dropWhile isSpace rest
  if null rest' then Right book else Left "Unexpected input"

natToBin :: Int -> String
natToBin n
  | n < 0 = error ("Invalid natural number: " ++ show n)
  | otherwise = replicate n '1' ++ "0"

bitsPerCount :: Int -> Int
bitsPerCount n = ceiling (logBase 2 (fromIntegral (max n 1)) :: Double)

bitsPerRef :: Book -> Int
bitsPerRef = bitsPerCount . length

indexToBits :: Int -> Int -> String
indexToBits index width
  | index < 0 || index >= 2 ^ width =
      error ("Invalid global reference index: " ++ show index)
  | otherwise =
      let bits = toBinary index
      in replicate (width - length bits) '0' ++ bits
  where
    toBinary 0 = ""
    toBinary n = toBinary (n `div` 2) ++ show (n `mod` 2)

refToBin :: String -> Book -> String
refToBin n defs = case findIndex ((== n) . fst) defs of
  Just i -> indexToBits i (bitsPerRef defs)
  Nothing -> error ("Undefined reference: @" ++ n)

termToBin :: Book -> Int -> Term -> String
termToBin defs d term = case term of
  Var n -> "10" ++ natToBin (d - read (drop 1 n) - 1)
  Ref n -> "11" ++ refToBin n defs
  Lam _ body -> "00" ++ termToBin defs (d + 1) (body (Var ("$" ++ show d)))
  App f x -> "01" ++ termToBin defs d f ++ termToBin defs d x

toBin :: Book -> String
toBin book = natToBin (length book) ++ concatMap (termToBin book 0 . snd) book

readNat :: String -> Either String (Int, String)
readNat bits =
  let (ones, rest) = span (== '1') bits
  in case rest of
    '0':xs -> Right (length ones, xs)
    _ -> Left "Invalid binary"

takeBits :: Int -> String -> Either String (String, String)
takeBits n bits =
  let (chunk, rest) = splitAt n bits
  in if length chunk == n && all (`elem` "01") chunk
    then Right (chunk, rest)
    else Left "Invalid binary"

bitsToIndex :: Int -> String -> Either String (Int, String)
bitsToIndex width bits
  | width == 0 = Right (0, bits)
  | otherwise = do
      (chunk, rest) <- takeBits width bits
      pure (foldl (\n c -> n * 2 + if c == '1' then 1 else 0) 0 chunk, rest)

readTerm :: Int -> [String] -> String -> Either String (String, String)
readTerm d defs bits = case bits of
  '1':'0':rest -> do
    (n, rest') <- readNat rest
    let i = d - n - 1
    if i >= 0 then pure (nameOf i, rest') else Left "Invalid binary"
  '1':'1':rest -> do
    (i, rest') <- bitsToIndex (bitsPerCount (length defs)) rest
    if i < length defs
      then pure ("@" ++ defs !! i, rest')
      else Left "Invalid binary"
  '0':'0':rest -> do
    let n = nameOf d
    (body, rest') <- readTerm (d + 1) defs rest
    pure ("\955" ++ n ++ "." ++ body, rest')
  '0':'1':rest -> readApp d defs rest
  _ -> Left "Invalid binary"

readApp :: Int -> [String] -> String -> Either String (String, String)
readApp d defs bits = do
  (f, rest) <- readTerm d defs bits
  (x, rest') <- readTerm d defs rest
  let head' = if "\955" `isPrefixOf` f then "(" ++ f ++ ")" else f
  pure (head' ++ "(" ++ x ++ ")", rest')

fromBin :: String -> Either String String
fromBin bits = do
  (count, rest) <- readNat bits
  let defs = [nameOf i | i <- [0 .. count - 1]]
  (lines', rest') <- go defs defs rest
  if null rest'
    then Right (unlines lines')
    else Left "Invalid binary"
  where
    go _ [] rest = Right ([], rest)
    go allDefs (d:ds) rest = do
      (term, rest') <- readTerm 0 allDefs rest
      (terms, rest'') <- go allDefs ds rest'
      pure (("@" ++ d ++ " = " ++ term) : terms, rest'')

mergeUses :: Uses -> Uses -> Uses
mergeUses xs ys = foldl add xs ys where
  add acc (k, n) = case lookup k acc of
    Nothing -> (k, n) : acc
    Just m -> (k, m + n) : filter ((/= k) . fst) acc

checkTerm :: String -> Int -> Term -> (Uses, [String])
checkTerm dn d term = case term of
  Var n -> ([(n, 1)], [])
  Ref _ -> ([], [])
  Lam _ body ->
    let v = "$" ++ show d
        (uses, errs) = checkTerm dn (d + 1) (body (Var v))
        count = maybe 0 id (lookup v uses)
        uses' = filter ((/= v) . fst) uses
        err = "@" ++ dn ++ ": \955" ++ nameOf d ++ " used "
          ++ show count ++ " times"
    in (uses', if count > 1 then err : errs else errs)
  App f x ->
    let (fu, fe) = checkTerm dn d f
        (xu, xe) = checkTerm dn d x
    in (mergeUses fu xu, fe ++ xe)

checkAffine :: Book -> [String]
checkAffine book = concatMap checkDef book where
  checkDef (n, term) = snd (checkTerm n 0 term)

wnf :: Book -> Term -> (Term, Int)
wnf book term = case term of
  Var _ -> (term, 0)
  Ref n -> case lookup n book of
    Nothing -> (term, 0)
    Just t -> wnf book t
  Lam _ _ -> (term, 0)
  App f x ->
    let (f', c) = wnf book f
    in case f' of
      Lam _ body ->
        let (r, c') = wnf book (body x)
        in (r, c + 1 + c')
      _ -> (App f' x, c)

snfAt :: Int -> Book -> Term -> (Term, Int)
snfAt d book term =
  let (term', c) = wnf book term
  in case term' of
    Var _ -> (term', c)
    Ref _ -> (term', c)
    Lam n body ->
      let (b, c') = snfAt (d + 1) book (body (Var (nameOf d)))
      in (Lam n (const b), c + c')
    App f x ->
      let (f', c1) = snfAt d book f
          (x', c2) = snfAt d book x
      in (App f' x', c + c1 + c2)

snf :: Book -> Term -> (Term, Int)
snf = snfAt 0
